### PR TITLE
Benny/store everything

### PR DESCRIPTION
**Pull request by benny-conn**
Changes:

- **Changed** token processing pipeline to always attempt to cache media (except http/iframes) and return errors where it is unable to.
- **Changed** media URL determination step to use explicitly defined cached objects and the URLs associated with them. Rather than going through a priority pipeline checking whether we have access to certain URLs, just use the URLs we know we have because the download/cache step returned the objects that were cached.
- **Removed** deleting of media. Because the media URLs used will be explicit, we no longer need to delete media to prevent it from being detected by the priority pipeline we previously used for determining which URLs to use in the media.
- **Changed** the naming scheme of objects to `{chain}-{contractAddress}-{tokenID}-{objectType}`
- **Added** object metadata that could be helpful in debugging such as which URL was thumbnailed to create a thumbnail or liverender and what the original URL and detected media type was for a cached object

TODO:

- [x] modify tests

## Release Notes
- New Feature: Modified token processing pipeline, media URL determination step, and naming scheme of objects in GraphQL.
- Bug Fix: Specific errors are now returned when there are DNS or URL errors in `GetDataFromURI` and `GetDataFromURIAsReader`.
- Refactor: Added a `String()` method to the `MediaType` type.
- Test: Modified the `MultiErr` struct's `Error()` method to only add non-nil errors to the error string.

> "Tokens processed with care,  
> Media URLs now aware.  
> Bugs squashed with specific flair,  
> Codebase clean beyond compare."

### CODE REVIEW

**Review by radazen — APPROVED** [review]
I'm so excited for this! I left some small questions/comments, but nothing too notable!

One other question: should we store the new stuff in a different GCS bucket to make it easier to delete all the old content once we're done with it?

File: service/media/media.go [hunk 1]
```diff
@@ -117,279 +153,201 @@ func NewStorageClient(ctx context.Context) *storage.Client {
...
+			if obj.objectType == ObjectTypeImage || obj.objectType == ObjectTypeSVG {
+				thumbnailObject = &obj
+			} else if obj.objectType == ObjectTypeThumbnail {
+				thumbnailObject = &obj
+				break
+			}
+		}
+	} else {
+		for _, obj := range objects {
+			if obj.objectType == ObjectTypeThumbnail {
```
[the remark this answers is not in the document]
Comment (benny-conn) [reply service/media/media.go:276]: Animation's should have a thumbnail and that thumbnail can be sort of non-explicit where it is just another provided image or svg, or it is labeled explicitly as a thumbnail. For image type objects, we don't need to use other image type objects to generate a thumbnail, they only get thumbnails when there was explicitly a thumbnail set by the creator of the token. I will make this more clear in comments

File: service/media/media.go [hunk 2]
```diff
@@ -117,279 +153,201 @@ func NewStorageClient(ctx context.Context) *storage.Client {
...
-		})
+func createMediaFromCachedObjects(ctx context.Context, tokenBucket string, objects []cachedMediaObject) persist.Media {
+	var primaryObject cachedMediaObject
+	for _, obj := range objects {
+		switch obj.objectType {
+		case ObjectTypeAnimation:
+			primaryObject = obj
+			break
+		case ObjectTypeImage, ObjectTypeSVG:
+			primaryObject = obj
```
[the remark this answers is not in the document]
Comment (benny-conn) [reply service/media/media.go:259]: The primary object should be the animation unless there is no animation, in which case it is just an image like object. That is why the animation has a break (stop if we find an animation) and the others don't. I will make this more clear in comments.

File: service/media/media.go [hunk 3]
```diff
@@ -756,96 +667,190 @@ func retryWriteToCloudStorage(ctx context.Context, writer io.Writer, reader io.R
...
-	return nil
+	return obj, nil
 }
 
-func createLiveRenderAndCache(ctx context.Context, videoURL, bucket, name string, client *storage.Client) error {
+func createLiveRenderAndCache(ctx context.Context, tids persist.TokenIdentifiers, videoURL, bucket string, client *storage.Client) (cachedMediaObject, error) {
+
+	obj := cachedMediaObject{
+		objectType:      ObjectTypeLiveRender,
+		mediaType:       persist.MediaTypeImage,
```
[the remark this answers is not in the document]
Comment (benny-conn) [reply service/media/media.go:829]: You are right, it should be video 👍

**Pull request merged**